### ISSUE TITLE
Test Explorer shows no tests if QUnit  is aliased inside a closure

### ISSUE THREAD

**Issue by jeroenheijmans** [mmanela/chutzpah#421]
I've started a fresh project and decided to shorthand QUnit in my test files. Here's an example of what  the code in `itemTests.js` looks like:

    (function (q) {
        q.module("Testmodule");
        q.test("Can create default item", function (assert) {
            assert.ok(!!new ConstructorFn());
        });
    }(QUnit));

The file being tested (`item.js`) is simply:

    var ConstructorFn = function () { };

The Test Explorer *did* show my tests, but that was merely because I had a `test.html` file that allows you to run all the tests without Visual Studio or Chutzpah. Some troubles with this include that double-clicking a test took me to that `test.html` file, not the test itself, and that I need to "touch" the `test.html` file before renames/deletions/additions to the set of tests are reflected in the Test Explorer.

So I removed the `test.html` file for testing purposes, but to my surprise all tests disappeared from the Test Explorer. There is no warning/indication of this in the Build or Test output, or any feedback from the Test Explorer on the reason my tests are missing. I searched with Google and Stack Overflow and didn't find anything either.

After some more debugging, it turned out that **this *does* work**:

    (function () {
        QUnit.module("Testmodule");
        QUnit.test("Can create default item", function (assert) {
            assert.ok(!!new ConstructorFn());
        });
    }());

In hindsight this kind-of makes sense, and if you know the internal workings of Chutzpah it probably makes a lot of sense.

If there's an easy fix, it would be fantastic if aliasing QUnit with an argument in the enclosing function is picked up by Chutpah.

If there's no easy fix or if it doesn't make sense for another reason to ignore this, that's fine: it's only a minor issue for me. Honestly, I mostly took the time spelling out this issue so others that make the same mistake can find this issue and see the workaround / solution here too.

Kind regards and thanks for creating / maintaining such a great piece of open source!

**Comment by mmanela** [comment]
There are a couple easy solutions to this. Chutzpah (unless you give it more info) guesses what your test runner framework is (since it supports QUnit, Jasmine and Mocha). In your case you broke the heuristic so Chutzpah won't run. There are two ways you can tell Chutzpah more explicitly that you are using QUnit:


1. Put a reference comment at the top of your file. (Don't worry qunit.js does not have to exist)
    
    ```/// <chutzpah_reference path="qunit.js" />```


2. Use a [chutzpah.json](https://github.com/mmanela/chutzpah/wiki/Chutzpah.json-Settings-File) to declare your framework.
```
{
    "Framework": "qunit"
}
```

**Comment by jeroenheijmans** [comment]
Both of your suggestions worked<sup>1</sup>, but in both cases double-clicking the test in the Test Explorer will only take me to the *file* with the test; it does not scroll to the appropriate test / set the cursor there (as it does with the workaround I mentioned in my first post).

----

<sup>1</sup> Option 1 did not work at first, because I placed that `chutzpah_reference` inside a file `myFixture.js` which I `reference`d from my tests file. Placing the `chutzpah_reference` inside the test file itself *did* work.

**Comment by mmanela** [comment]
Good! To make the line matching to work again you need to tell Chutzpah what your test method looks like. Out of the box Chutzpah looks for standard ways defined in QUnit but if you customize you just just give Chutzpah a regex in the [TestPattern setting](https://github.com/mmanela/chutzpah/wiki/testpattern-setting)

**Comment by jeroenheijmans** [comment]
With that I got everything working again. For completeness sake, as well as others that may find this issue while Googling, here's the changed settings I ended up with:

	{
		"Framework": "qunit",
		"TestPattern": "((?<!\\.)\\b(?:(q|QUnit)\\.)?(test|asyncTest)[\\t ]*\\([\\t ]*[\"'](?<TestName>.*)[\"'])"
	}

Notice that the change from the default bit is that `QUnit` is now `(q|QUnit)`.

With this `chutzpah.json` file I was able to run QUnit tests with Chutzpah in the Visual Studio Test Explorer once again, and double-clicking the tests takes me to the proper test.

Thanks @mmanela for some excellent and quick help. All this now makes sense, and things work as I'd expect. Now that I see *how* all this works I'm not sure if I'm going to stick with a custom chutzpah.json, because it's not very discoverable for people new to the codebase (e.g. they might choose a different shorthand variable name like `qu`, and be left wondering why their tests aren't showing), but that's fine by me.

Important footnote: don't forget (as I did at first) that this is JSON, and that the regex needs to have all characters properly escaped. At first I forgot to escape the slashes, which made my devenv crash.
